Updates for v0.6.0 (#56)
shipwright-io/website: content/en/blog/releases/v0.6.0.md

---
title: "Build v0.6.0"
date: 2021-10-11T16:00:00-04:00
draft: false
author: "Sascha Schwarze ([@SaschaSchwarze0](https://github.com/SaschaSchwarze0))"
---

> It was a long journey, a whole summer actually, and more than it should have been taken, but finally, here it is: v0.6.0 of Shipwright, our first release after having joined the [Continuous Delivery Foundation](https://cd.foundation/).

Reviewer: ```suggestion
It was a long journey, longer than it should have been, but we have finally reached our goal! Shipwright Build v0.6.0, our first release after joining the [Continuous Delivery Foundation](https://cd.foundation/).
```

> The feature list is impressive. It made it hard to pick some top 3. Here is my choice:

Reviewer: ```suggestion
It was hard to choose just three features from an impressive list, but here are my choices:
```

## Parameterization of Builds

> Our API objects; the build strategy that defines the steps to turn source code into a container image, the build where a user configures his container image build, and the build run as an instance of a build; so far suffered of a major capability: parameterization. We defined the available inputs that a user specified in a build. Those were only the minimum: the location of the source code, the name of the strategy and the destination image. Any other customization of the build was not possible, limiting the possible scenarios.

Reviewer: ```suggestion
Our API objects; the build strategy that defines the steps to turn source code into a container image, the build where a user configures their container image build, and the build run as an instance of a build; were missing a major capability: parameterization. We defined the available inputs that a user specified in a build. Those were only the minimum: the location of the source code, the name of the strategy and the destination image. Any other customization of the build was not possible, limiting the possible scenarios.
```

Two major enhancements bring us a huge step forward:

> 

Reviewer: Add a `Parameters` heading.

Build strategy authors can now define parameters that they can use then in their steps. Build users provide the values for those parameters. Our sample build strategies start to make use of this feature: the [BuildKit sample](https://github.com/shipwright-io/build/blob/v0.6.0/samples/buildstrategy/buildkit/buildstrategy_buildkit_cr.yaml#L14-L16) now contains a parameter to enable the usage of insecure container registry, and the [ko sample](https://github.com/shipwright-io/build/blob/v0.6.0/samples/buildstrategy/ko/buildstrategy_ko_cr.yaml#L8-L22) comes with a rich set of parameters to select Go version and flags, ko version and the main package directory. The [ko sample build](https://github.com/shipwright-io/build/blob/v0.6.0/samples/build/build_ko_cr.yaml#L10-L15) outlines how those parameters can be used to build the Shipwright Build controller image. Read more about parameters in the documentation for [build strategy authors](https://github.com/shipwright-io/build/blob/v0.6.0/docs/buildstrategies.md#strategy-parameters) and [build users](https://github.com/shipwright-io/build/blob/v0.6.0/docs/build.md#defining-paramvalues).

> 

Reviewer: Add an `Environment Variables` heading.

> Build strategy authors can now define parameters that they can use then in their steps. Build users provide the values for those parameters. Our sample build strategies start to make use of this feature: the [BuildKit sample](https://github.com/shipwright-io/build/blob/v0.6.0/samples/buildstrategy/buildkit/buildstrategy_buildkit_cr.yaml#L14-L16) now contains a parameter to enable the usage of insecure container registry, and the [ko sample](https://github.com/shipwright-io/build/blob/v0.6.0/samples/buildstrategy/ko/buildstrategy_ko_cr.yaml#L8-L22) comes with a rich set of parameters to select Go version and flags, ko version and the main package directory. The [ko sample build](https://github.com/shipwright-io/build/blob/v0.6.0/samples/build/build_ko_cr.yaml#L10-L15) outlines how those parameters can be used to build the Shipwright Build controller image. Read more about parameters in the documentation for [build strategy authors](https://github.com/shipwright-io/build/blob/v0.6.0/docs/buildstrategies.md#strategy-parameters) and [build users](https://github.com/shipwright-io/build/blob/v0.6.0/docs/build.md#defining-paramvalues).

Reviewer: It would probably be better to link to just the yaml file, instead of specific line numbers, as those will change over time.

Reviewer: This links to line numbers: `[BuildKit sample](https://github.com/shipwright-io/build/blob/v0.6.0/samples/buildstrategy/buildkit/buildstrategy_buildkit_cr.yaml#L14-L16)`

Author: Right, sry, better explanation: I think it is fine to link to the line number because I referenced the v0.6.0 tag which will never change. We definitely should not link to line numbers in main, but that's not happening here.

> But, not in all scenarios, it is possible to define all possible parameters upfront in a build strategy. For those tools that control behavior through environment variables, users can now define environment variables in their builds and build runs. Environment variable values can be referenced from Kubernetes secrets to support scenarios where for example tokens for private dependency management systems are needed. Examples of builds and build runs with environment variables can be found in our [build](https://github.com/shipwright-io/build/blob/v0.6.0/docs/build.md) and [build run](https://github.com/shipwright-io/build/blob/v0.6.0/docs/buildrun.md) documentation.

Reviewer: ```suggestion
It is not always possible to define all of the parameters upfront in a build strategy. For those tools that control behavior through environment variables, users can now define environment variables in their builds and build runs. Environment variable values can be referenced from Kubernetes secrets to support scenarios where for example tokens for private dependency management systems are needed. Examples of builds and build runs with environment variables can be found in our [build](https://github.com/shipwright-io/build/blob/v0.6.0/docs/build.md) and [build run](https://github.com/shipwright-io/build/blob/v0.6.0/docs/buildrun.md) documentation.
```

## CLI

> Being based on Kubernetes, all our objects are Kubernetes custom resources with their ecosystem-specific syntax and concepts. But, not all our users should require a black belt in using Kubernetes. We are therefore proud to release the first version of our CLI. Build users can use it to easily manage their builds and build runs. Some highlights:

Reviewer: ```suggestion
Being based on Kubernetes, all our objects are Kubernetes custom resources with their ecosystem-specific syntax and concepts. But, our users should not be required to be experts in using Kubernetes. We are therefore proud to release the first version of our CLI. Build users can use it to easily manage their builds and build runs. Some highlights:
```

Reviewer: It is only a suggestion, I usually shy away from metaphors and local colloquialisms as not everyone understands their meaning.

Assuming you are connected to your Kubernetes cluster and have the credentials for your container registry in place, then creating a build for our go-sample is a single command:

```sh
$ shp build create sample-go --source-url https://github.com/shipwright-io/sample-go --source-context-dir source-build --output-image <CONTAINER_REGISTRY_LOCATION>/sample-go --output-credentials-secret <CONTAINER_REGISTRY_SECRET>
Created build "sample-go"
```

> And then it is just another single command to trigger a run and and to follow its logs. Nice. 😊

Reviewer: ```suggestion
And then it is just another single command to trigger a build run and and to follow its logs. Nice. 😊
```

```sh
$ shp build run sample-go --follow
Pod 'sample-go-9hbpj-l9t5j-pod-fz2sp' is in state "Pending"...
Pod 'sample-go-9hbpj-l9t5j-pod-fz2sp' is in state "Pending"...
[source-default] 2021/10/08 11:43:33 Info: ssh (/usr/bin/ssh): OpenSSH_8.0p1, OpenSSL 1.1.1g FIPS  21 Apr 2020
[source-default] 2021/10/08 11:43:33 Info: git (/usr/bin/git): git version 2.27.0
[source-default] 2021/10/08 11:43:33 Info: git-lfs (/usr/bin/git-lfs): git-lfs/2.11.0 (GitHub; linux arm64; go 1.14.4)
[source-default] 2021/10/08 11:43:33 /usr/bin/git clone -h
[source-default] 2021/10/08 11:43:33 /usr/bin/git clone --quiet --no-tags --single-branch --depth 1 -- https://github.com/shipwright-io/sample-go /workspace/source
...
```

> 

Reviewer: Should we include an example of using the `shp buildrun create` command here also?

> So far so good. Please keep in mind that this is our first CLI release. We know that some of the new backend features are not yet available through it, and the command layout will likely go through the one or other iteration until we are all happy with the usability and consistency.

Reviewer: ```suggestion
So far so good. Please keep in mind that this is our first CLI release. We know that some of the new backend features are not yet available through it, and the command layout will likely go through subsequent iterations until we are all happy with the usability and consistency.
```

## Post-processing the image

So far, the images that were produced in the strategy's tools were untouched. Now we started going beyond this:

> A community-provided extension for the Kaniko build strategy sample provides image scanning using the [Trivi tool](https://github.com/aquasecurity/trivy). This is a show case on how one can provide enhanced capabilities even as build strategy author. Read about [scanning with Trivy](https://github.com/shipwright-io/build/blob/v0.6.0/docs/buildstrategies.md#scanning-with-trivy) and check out how the [build strategy has been built](https://github.com/shipwright-io/build/blob/v0.6.0/samples/buildstrategy/kaniko/buildstrategy_kaniko-trivy_cr.yaml).

Reviewer: ```suggestion
A community-provided extension for the Kaniko build strategy sample provides image scanning using the [Trivi tool](https://github.com/aquasecurity/trivy). This is an example of how a build strategy author can provide advanced capabilities. Read about [scanning with Trivy](https://github.com/shipwright-io/build/blob/v0.6.0/docs/buildstrategies.md#scanning-with-trivy) and check out how the [build strategy has been built](https://github.com/shipwright-io/build/blob/v0.6.0/samples/buildstrategy/kaniko/buildstrategy_kaniko-trivy_cr.yaml).
```

> Another feature provides first-class support for labels and annotations that build users can provide. Those are added to the image after the build strategy steps are done. Read more in [Defining the output of a build](https://github.com/shipwright-io/build/blob/v0.6.0/docs/build.md#defining-the-output).

Reviewer: ```suggestion
Another feature enables first-class support for labels and annotations provided by build users. Those are added to the image after the build strategy steps are done. Read more in [Defining the output of a build](https://github.com/shipwright-io/build/blob/v0.6.0/docs/build.md#defining-the-output).
```

--

> A lot nice stuff, is that all ? No. 😊 Some other small but still nice features:

Reviewer: ```suggestion
A lot of nice stuff, but is that all ? No. 😊 Some other small but still nice features:
```

> * You can now cancel a running build run which is very helpful if you just triggered it and then – life is life – look back at your source code and know that you'll have to push another change making the running build valueless. TODO link to doc

Reviewer: ```suggestion
* You can now cancel a running build run which is very helpful if you just triggered it and then look back at your source code and know that you'll have to push another change making the running build useless. TODO link to doc
```

> * The build run status has been improved: failed build runs have an improved status. Successful build runs now contain metadata about the source (sha and author of the latest commit) and the image that was build (digest and size, depending on the build strategy author to provide them). TODO link to doc

Reviewer: ```suggestion
* The build run status has been improved: failed build runs have an improved status. Successful build runs now contain metadata about the source (sha and author of the latest commit) and the image that was built (digest and size, depending on the build strategy author to provide them). TODO link to doc
```

* We improved our build process so that we finally have release-specific sample build strategies.

> And maybe there's more that I am now overseeing.

Reviewer: ```suggestion
For a full list of updates you can check out the release [change log](link here).
```

> If thatNow, how to get started? Assuming your Kubernetes cluster ready, it’s all done with just these four commands, to install Tekton, Shipwright, the sample build strategies, and the CLI.

Reviewer: ```suggestion
Now, how to get started? Assuming your Kubernetes cluster ready, it’s all done with just these four commands, which install Tekton, Shipwright, the sample build strategies, and the CLI.
```

> Now, how to get started? Assuming your Kubernetes cluster ready, it’s all done with just these four commands, which install Tekton, Shipwright, the sample build strategies, and the CLI.

Reviewer: ```suggestion
Now, how to get started? Assuming your Kubernetes cluster is ready, it’s all done with just these four commands, which install Tekton, Shipwright, the sample build strategies, and the CLI.
```

```sh
kubectl apply -f https://storage.googleapis.com/tekton-releases/pipeline/previous/v0.25.0/release.yaml
kubectl apply -f https://github.com/shipwright-io/build/releases/download/v0.6.0/release.yaml
kubectl apply -f https://github.com/shipwright-io/build/releases/download/v0.6.0/sample-strategies.yaml

> TODO command for CLI installation

Reviewer 2: Noting this TODO here (to be fixed once we have CLI binaries on github)

---

> title: "Build v0.6.0"

Author: I originally also planned to use a more fancy title, but as it is the blog post for a release, it will appear inside the [Releases category](https://shipwright.io/blog/releases/) where I am continuing the naming the way it is for all previous posts.

date: 2021-10-11T16:00:00-04:00
draft: false
author: "Sascha Schwarze ([@SaschaSchwarze0](https://github.com/SaschaSchwarze0))"
---

> It was a long journey, longer than it should have been, but we have finally reached our goal! Shipwright Build v0.6.0, our first release after joining the [Continuous Delivery Foundation](https://cd.foundation/).

Reviewer 2: Our prior release announcements tended to be more terse - mainly pointing to the release notes on GitHub.

This page should have the filename `v0.6.0-release-notes.md`, and highlight the breaking changes/upgrade instructions. For some reason the post is not rendering - are you able to build the site locally with Hugo?

Perhaps move this content to a different file the `posts/` directory, and give it a filename that starts with the blog post date? Ex - `2021-10-11-build-release-v0.6.0`
